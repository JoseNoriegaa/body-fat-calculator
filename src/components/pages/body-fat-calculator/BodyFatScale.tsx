// External dependencies
import React, { useMemo, useState, useRef, useLayoutEffect } from 'react';

// Internal dependencies
import { BODY_FAT_CATEGORIES } from '@/constants';
import useStyles from './BodyFatScale.styles';

// Types & Interfaces
interface IBodyFatScaleProps {
  bodyFatPercentage: number;
  isMale: boolean;
}

function BodyFatScale({ bodyFatPercentage, isMale }: IBodyFatScaleProps) {
  const scaleRef = useRef<HTMLDivElement>(null);
  const [scaleComponentWidth, setScaleComponentWidth] = useState<number>(0);

  const classes = useStyles();

  useLayoutEffect(() => {
    if (scaleRef.current) {
      setScaleComponentWidth(scaleRef.current.offsetWidth);
    }
  }, []);

  const pointerPosition = useMemo(() => {
    const percentage = Math.min((bodyFatPercentage * 100) / 33, 100);

    if (scaleComponentWidth * (percentage / 100) < 60) {
      return '60px';
    }

    return `${percentage}%`;
  }, [bodyFatPercentage, scaleComponentWidth]);

  const categories = isMale ? BODY_FAT_CATEGORIES.MALE : BODY_FAT_CATEGORIES.FEMALE;

  return (
    <>
      <h3 className={classes.title}>Tu resultado: {bodyFatPercentage.toFixed(1)}%</h3>

      <div className={classes.scaleOuterContainer}>
        <div className={classes.pointerIndicator} style={{ left: pointerPosition }}>
          <span className={classes.percentage}>{bodyFatPercentage.toFixed(1)}%</span>
          <div className={classes.pointerIndicatorIcon} />
        </div>

        <div ref={scaleRef} className={classes.scaleContainer} />

        <div className={classes.indicators}>
          {categories.map((category) => (
            <div className={classes.indicatorContainer} key={category.description}>
              <div className={classes.indicatorColor} />

              <span className={classes.indicatorPercentage}>{category.value}</span>
              <p className={classes.indicatorText}>{category.description}</p>
            </div>
          ))}
        </div>
      </div>
    </>
  );
}

export default BodyFatScale;
