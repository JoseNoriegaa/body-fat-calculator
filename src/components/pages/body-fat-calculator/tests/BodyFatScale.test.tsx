// External dependencies
import { render, screen } from '@testing-library/react';

// Internal dependencies
import BodyFatScale from '../BodyFatScale';

describe('<BodyFatScale />', () => {
  it.each(['male', 'female'])('should match snapshot - male', (gender) => {
    const { asFragment } = render(
      <BodyFatScale isMale={gender === 'male'} bodyFatPercentage={20} />
    );

    expect(screen.getByText('Tu resultado: 20.0%')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
