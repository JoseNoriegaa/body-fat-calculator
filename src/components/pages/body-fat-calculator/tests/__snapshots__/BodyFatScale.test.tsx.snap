// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BodyFatScale /> should match snapshot - male 1`] = `
<DocumentFragment>
  <h3
    class="title-0-2-1"
  >
    Tu resultado: 20.0%
  </h3>
  <div
    class="scaleOuterContainer-0-2-2"
  >
    <div
      class="pointerIndicator-0-2-3"
      style="left: 60px;"
    >
      <span
        class="percentage-0-2-4"
      >
        20.0%
      </span>
      <div
        class="pointerIndicatorIcon-0-2-5"
      />
    </div>
    <div
      class="scaleContainer-0-2-6"
    />
    <div
      class="indicators-0-2-8"
    >
      <div
        class="indicatorContainer-0-2-7"
      >
        <div
          class="indicatorColor-0-2-9"
        />
        <span
          class="indicatorPercentage-0-2-10"
        >
          2-4%
        </span>
        <p
          class="indicatorText-0-2-11"
        >
          Esencial
        </p>
      </div>
      <div
        class="indicatorContainer-0-2-7"
      >
        <div
          class="indicatorColor-0-2-9"
        />
        <span
          class="indicatorPercentage-0-2-10"
        >
          6-13%
        </span>
        <p
          class="indicatorText-0-2-11"
        >
          Deportista
        </p>
      </div>
      <div
        class="indicatorContainer-0-2-7"
      >
        <div
          class="indicatorColor-0-2-9"
        />
        <span
          class="indicatorPercentage-0-2-10"
        >
          14-17%
        </span>
        <p
          class="indicatorText-0-2-11"
        >
          Fitness
        </p>
      </div>
      <div
        class="indicatorContainer-0-2-7"
      >
        <div
          class="indicatorColor-0-2-9"
        />
        <span
          class="indicatorPercentage-0-2-10"
        >
          18-25%
        </span>
        <p
          class="indicatorText-0-2-11"
        >
          Aceptable
        </p>
      </div>
      <div
        class="indicatorContainer-0-2-7"
      >
        <div
          class="indicatorColor-0-2-9"
        />
        <span
          class="indicatorPercentage-0-2-10"
        >
          25%
        </span>
        <p
          class="indicatorText-0-2-11"
        >
          Obeso
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<BodyFatScale /> should match snapshot - male 2`] = `
<DocumentFragment>
  <h3
    class="title-0-2-1"
  >
    Tu resultado: 20.0%
  </h3>
  <div
    class="scaleOuterContainer-0-2-2"
  >
    <div
      class="pointerIndicator-0-2-3"
      style="left: 60px;"
    >
      <span
        class="percentage-0-2-4"
      >
        20.0%
      </span>
      <div
        class="pointerIndicatorIcon-0-2-5"
      />
    </div>
    <div
      class="scaleContainer-0-2-6"
    />
    <div
      class="indicators-0-2-8"
    >
      <div
        class="indicatorContainer-0-2-7"
      >
        <div
          class="indicatorColor-0-2-9"
        />
        <span
          class="indicatorPercentage-0-2-10"
        >
          10-13%
        </span>
        <p
          class="indicatorText-0-2-11"
        >
          Esencial
        </p>
      </div>
      <div
        class="indicatorContainer-0-2-7"
      >
        <div
          class="indicatorColor-0-2-9"
        />
        <span
          class="indicatorPercentage-0-2-10"
        >
          14-20%
        </span>
        <p
          class="indicatorText-0-2-11"
        >
          Deportista
        </p>
      </div>
      <div
        class="indicatorContainer-0-2-7"
      >
        <div
          class="indicatorColor-0-2-9"
        />
        <span
          class="indicatorPercentage-0-2-10"
        >
          21-24%
        </span>
        <p
          class="indicatorText-0-2-11"
        >
          Fitness
        </p>
      </div>
      <div
        class="indicatorContainer-0-2-7"
      >
        <div
          class="indicatorColor-0-2-9"
        />
        <span
          class="indicatorPercentage-0-2-10"
        >
          25-31%
        </span>
        <p
          class="indicatorText-0-2-11"
        >
          Aceptable
        </p>
      </div>
      <div
        class="indicatorContainer-0-2-7"
      >
        <div
          class="indicatorColor-0-2-9"
        />
        <span
          class="indicatorPercentage-0-2-10"
        >
          32%
        </span>
        <p
          class="indicatorText-0-2-11"
        >
          Obeso
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
